### PR TITLE
Feat/export csv exchange

### PR DESCRIPTION
**Pull request by varun257**
### Description of the Change

export groups to csv
export external contacts to csv
export shared accounts to csv

### Benefits

Export option for consumers

### Possible Drawbacks

None

### Applicable Issues

MBP-23

### CODE REVIEW

File: src/exchange/external-contact/external-contact.html [hunk 1]
```diff
@@ -122,6 +122,11 @@
                     data-translate="exchange_tab_EXTERNAL_CONTACTS_configuration_contact_add_title_button"
                     data-ng-click="setAction('exchange/external-contact/add/external-contact-add')">
             </button>
+            <button class="btn btn-block btn-default"
+                    type="button"
+                    data-ng-click="ctrl.services.navigation.setAction('exchange/account/export-as-csv/account-export-as-csv', { search : ctrl.filter, total : ctrl.contacts.list.results.length, exchange : ctrl.exchange, csvExportType: 'external' })"
+                    data-ng-disabled="ctrl.mailingLists.list.results.count === 0"
+                    data-ng-bind=":: 'exchange_tab_EXTERNAL_CONTACTS_export_csv' | translate"></button>
```
Comment (marie-j) [inline src/exchange/external-contact/external-contact.html]:
```suggestion
                    data-translate=":exchange_tab_EXTERNAL_CONTACTS_export_csv"></button>
```

File: src/exchange/group/group.html [hunk 2]
```diff
@@ -162,6 +162,11 @@
                     data-translate="exchange_tab_GROUPS_all_add_ml_button"
                     data-ng-click="ctrl.newGroup()">
             </button>
+            <button class="btn btn-block btn-default"
+                    type="button"
+                    data-ng-click="ctrl.services.navigation.setAction('exchange/account/export-as-csv/account-export-as-csv', { search : ctrl.search.value, total : ctrl.mailingLists.list.results.length, exchange : ctrl.exchange, csvExportType: 'group' })"
+                    data-ng-disabled="ctrl.mailingLists.list.results.count === 0"
+                    data-ng-bind=":: 'exchange_GRUOUPS_accounts_export_csv' | translate"></button>
```
Comment (marie-j) [inline src/exchange/group/group.html]:
```suggestion
                    data-translate="exchange_GRUOUPS_accounts_export_csv"></button>
```

File: src/exchange/shared-account/shared-account.html [hunk 3]
```diff
@@ -144,6 +144,11 @@
                     data-ng-click="ctrl.newSharedAccount()"
                     data-ng-if="ctrl.isSharedAccountAdjustable()">
             </button>
+            <button class="btn btn-block btn-default"
+                    type="button"
+                    data-ng-click="ctrl.services.navigation.setAction('exchange/account/export-as-csv/account-export-as-csv', { search : ctrl.searchValue, total : ctrl.accounts.list.results.length, exchange : ctrl.exchange, csvExportType: 'shared' })"
+                    data-ng-disabled="ctrl.mailingLists.list.results.count === 0"
+                    data-ng-bind=":: 'exchange_SHARED_ACCOUNTS_export_csv' | translate"></button>
```
Comment (marie-j) [inline src/exchange/shared-account/shared-account.html]:
```suggestion
                    data-translate="exchange_SHARED_ACCOUNTS_export_csv"></button>
```

File: src/exchange/external-contact/external-contact.html [hunk 4]
```diff
@@ -122,6 +122,11 @@
                     data-translate="exchange_tab_EXTERNAL_CONTACTS_configuration_contact_add_title_button"
                     data-ng-click="setAction('exchange/external-contact/add/external-contact-add')">
             </button>
+            <button class="btn btn-block btn-default"
```
[the remark this answers is not in the document]
Comment (varun257) [reply src/exchange/external-contact/external-contact.html]: have added `oui-kit` classes

**Comment by varun257** [comment]
@marie-j Have taken care of all comments

File: src/exchange/account/export-as-csv/account-export-as-csv.controller.js [hunk 5]
```diff
@@ -137,17 +163,55 @@ angular.module('Module.exchange.controllers').controller(
...
 
     prepareForCsv(exportOpts, offset, infos, timeoutObject) {
-      return this.services.Exchange.prepareForCsv(
-        this.$routerParams.organization,
-        this.$routerParams.productId,
-        exportOpts,
-        offset,
-        timeoutObject,
-      ).then((datas) => {
+      this.promise = null;
```
Comment (antleblanc) [inline src/exchange/account/export-as-csv/account-export-as-csv.controller.js]:
```suggestion
      let promise = null;
```

File: src/exchange/exchange.service.js [hunk 6]
```diff
@@ -805,6 +805,110 @@ angular.module('Module.exchange.services').service(
...
+                  angular.forEach(aliases.list.results, (alias) => {
+                    account.aliases.push(alias.displayName);
+                  });
+                }),
+              );
+            } else {
+              _.set(account, 'aliases', []);
+            }
+            if (account.managers > 0) {
+              _.set(account, 'managers', []);
```
[the remark this answers is not in the document]
Comment (varun257) [reply src/exchange/exchange.service.js]: function parameters cannot be assigned directly. I get eslint error if i do that

File: src/exchange/exchange.service.js [hunk 7]
```diff
@@ -805,6 +805,110 @@ angular.module('Module.exchange.services').service(
...
+              queue.push(
+                this.getGroupManagerList(
+                  organization,
+                  serviceName,
+                  account.mailingListAddress,
+                ).then((managers) => {
+                  angular.forEach(managers.list.results, (manager) => {
```
[the remark this answers is not in the document]
Comment (varun257) [reply src/exchange/exchange.service.js]: I cannot assign to `account.managers` directly by doing `account.managers = angular.map()` results in eslint error

[the remark this answers is not in the document]
Comment (varun257) [reply src/exchange/exchange.service.js]: Have used map with `_.set()`. please have a look now

**Pull request merged**